File: 16.最接近的三数之和.go


/*
 * @lc app=leetcode.cn id=16 lang=golang
 *
 * [16] 最接近的三数之和
 */

import (
	"sort"
)

// @lc code=start

// 获取绝对值
func abs(num int) int {
	if num < 0 {
		return num * -1
	}
	return num
}

func threeSumClosest(nums []int, target int) int {
	var n = len(nums) - 1
	var sum = nums[0] + nums[1] + nums[2]

	// 将数组进行排序
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})

	// 遍p点
	for p := 0; p < n-1; p += 1 {
		// 求出差值
		anoter := target - nums[p]
		// 赋低、矮两个指针
		l, h := p+1, n
		for l < h {
			if l == p || h == p {
				continue
			}
			v := nums[l] + nums[h]
			w := v + nums[p]
			if abs(target-w) < abs(target-sum) {
				sum = w
			}
			if v > anoter {
				h--
			} else if v < anoter {
				l++
			} else {
				break
			}
		}
	}
	return sum
}

// @lc code=end

// func main() {
// 	threeSumClosest([]int{-1, 2, 1, -4}, 2) //2
// }

File: 16.最接近的三数之和.ts
/*
 * @lc app=leetcode.cn id=16 lang=typescript
 *
 * [16] 最接近的三数之和
 */

// @lc code=start

function getSum(nums: number[], target: number): number {
  const sum = nums.reduce((x, y) => x + y);
  const res = Math.abs(target - sum);
  return res;
}

function threeSumClosest(nums: number[], target: number): number {
  let res: number = Number.MAX_VALUE;
  let x = 0, y = 0, z = 0;
  let sum: number = 0;
  const n = nums.length - 1;
  for (let p = 0; p < n; p++) {
    for (let i = p - 1; i >= 0; i -= 1) {
      for (let j = p + 1; j <= n; j += 1) {
        let a = nums[i];
        let b = nums[p];
        let c = nums[j];
        const _res = getSum([a, b, c], target);
        if (_res < res) {
          res = _res;
          x = i;
          y = p;
          z = j;
        }
      }
    }
  }
  sum = nums[x] + nums[y] + nums[z];
  return sum;
};
// @lc code=end

// threeSumClosest([-1,2,1,-4], 1)
// threeSumClosest([-3, -2, -5, 3, -4], -1) // -2
// threeSumClosest([1,1,1,0], -100) // 2
threeSumClosest([1, 2, 4, 8, 16, 32, 64, 128], 82)